source/simulator/solver.cc

```diff
@@ -841,26 +868,32 @@ namespace aspect
 
         solver_control_cheap.enable_history_data();
         solver_control_expensive.enable_history_data();
+        internal::InverseWeightedMassMatrix<TrilinosWrappers::PreconditionBase> inverse_weighted_mass_matrix(
+          system_preconditioner_matrix.block(1,1),
+          *Mp_preconditioner,
+          parameters.linear_solver_S_block_tolerance);
 
         // create a cheap preconditioner that consists of only a single V-cycle
         const internal::BlockSchurPreconditioner<LinearAlgebra::PreconditionAMG,
-              LinearAlgebra::PreconditionBase>
-              preconditioner_cheap (system_matrix, system_preconditioner_matrix,
-                                    *Mp_preconditioner, *Amg_preconditioner,
+              internal::InverseWeightedMassMatrix<TrilinosWrappers::PreconditionBase>>
+              preconditioner_cheap (system_matrix,
+                                    system_preconditioner_matrix,
+                                    inverse_weighted_mass_matrix,
+                                    *Amg_preconditioner,
                                     /* do_solve_A = */ false,
                                     stokes_A_block_is_symmetric(),
-                                    parameters.linear_solver_A_block_tolerance,
-                                    parameters.linear_solver_S_block_tolerance);
+                                    parameters.linear_solver_A_block_tolerance);
 
         // create an expensive preconditioner that solves for the A block with CG
         const internal::BlockSchurPreconditioner<LinearAlgebra::PreconditionAMG,
-              LinearAlgebra::PreconditionBase>
-              preconditioner_expensive (system_matrix, system_preconditioner_matrix,
-                                        *Mp_preconditioner, *Amg_preconditioner,
+              internal::InverseWeightedMassMatrix<TrilinosWrappers::PreconditionBase>>
+              preconditioner_expensive (system_matrix,
+                                        system_preconditioner_matrix,
+                                        inverse_weighted_mass_matrix,
```

Comment:
I think the issue is that you are reusing ``inverse_weighted_mass_matrix`` and when we switch from cheap to expensive, we no longer reset the number of s_iterations but add to it.